lib/plugin/neovim-plugin.nix

```diff
@@ -0,0 +1,102 @@
+{
+  lib,
+  nixvimOptions,
+  toLuaObject,
+  nixvimUtils,
+  mkPlugin,
+}:
+with lib;
+rec {
+  mkSettingsOption =
+    {
+      pluginName ? null,
+      options ? { },
+      description ?
+        if pluginName != null then
+          "Options provided to the `require('${pluginName}').setup` function."
+        else
+          throw "mkSettingsOption: Please provide either a `pluginName` or `description`.",
+      example ? null,
+    }:
+    nixvimOptions.mkSettingsOption { inherit options description example; };
+
+  # TODO: DEPRECATED: use the `settings` option instead
+  extraOptionsOptions = {
+    extraOptions = mkOption {
+      default = { };
+      type = with types; attrsOf anything;
+      description = ''
+        These attributes will be added to the table parameter for the setup function.
+        Typically, it can override NixVim's default settings.
+      '';
+    };
+  };
+
+  mkNeovimPlugin =
+    config:
+    {
+      name,
+      defaultPackage, # make this parameter mandatory for neovim plugins
+      # deprecations
+      deprecateExtraOptions ? false,
+      # colorscheme
+      isColorscheme ? false,
+      # options
+      settingsOptions ? { },
+      settingsExample ? null,
+      # config
+      luaName ? name,
+      callSetup ? true,
+      ...
+    }@args:
+    mkPlugin config (
+      (removeAttrs args [
+        "callSetup"
+        "deprecateExtraOptions"
+        "extraConfig"
+        "extraOptions"
+        "imports"
+        "luaName"
+        "settingsExample"
+        "settingsOptions"
+      ])
+      // {
+        imports =
+          let
+            basePluginPath = [
+              (if isColorscheme then "colorschemes" else "plugins")
+              name
+            ];
+            settingsPath = basePluginPath ++ [ "settings" ];
+          in
+          (args.imports or [ ])
+          ++ (optional deprecateExtraOptions (
+            mkRenamedOptionModule (basePluginPath ++ [ "extraOptions" ]) settingsPath
+          ));
+
+        extraOptions = {
+          settings = mkSettingsOption {
+            pluginName = name;
+            options = settingsOptions;
+            example = settingsExample;
+          };
+        } // (args.extraOptions or { });
+
+        extraConfig =
+          let
+            extraConfigNamespace = if isColorscheme then "extraConfigLuaPre" else "extraConfigLua";
```

Comment:
Make this a default arg instead?

`extraConfigNamespace ? if isColorscheme then "extraConfigLuaPre" else "extraConfigLua",`

Reply:
Ok